src/kern/npf_alg_pptp.c

```diff
@@ -0,0 +1,763 @@
+		alg_arg->gre_cons[i].flags = PPTP_ALG_FL_FREE_ENTRY;
+
+	return alg_arg;
+}
+
+/*
+ *
+ */
+static struct pptp_gre_con *
+npfa_pptp_alg_arg_lookup(struct pptp_alg_arg *arg)
+{
+	int i;
+
+	for (i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++)
+		if (arg->gre_cons[i].flags & PPTP_ALG_FL_FREE_ENTRY)
+			return &arg->gre_cons[i];
+
+	return NULL;
+}
+
+/*
+ *
+ */
+static struct pptp_gre_con *
+npfa_pptp_alg_arg_lookup_with_server_call_id(struct pptp_alg_arg *arg,
+		  uint16_t server_call_id)
+{
+	int i;
+	struct pptp_gre_con *gre_con;
+
+	for (i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++) {
+		gre_con = &arg->gre_cons[i];
+
+		if ((gre_con->flags & PPTP_ALG_FL_FREE_ENTRY) == 0 &&
+				  gre_con->ctx.server_call_id == server_call_id)
+			return gre_con;
+	}
+
+	return NULL;
+}
+
+/*
+ *
+ */
+static struct pptp_gre_con *
+npfa_pptp_alg_arg_lookup_with_orig_client_call_id(struct pptp_alg_arg *arg,
+		  uint16_t client_call_id)
+{
+	int i;
+	struct pptp_gre_con *gre_con;
+
+	for (i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++) {
+		gre_con = &arg->gre_cons[i];
+
+		if ((gre_con->flags & PPTP_ALG_FL_FREE_ENTRY) == 0 &&
+				  gre_con->orig_client_call_id == client_call_id)
+			return gre_con;
+	}
+
+	return NULL;
+}
+
+/*
+ *
+ */
+static struct pptp_gre_con *
+npfa_pptp_alg_arg_lookup_with_client_call_id(struct pptp_alg_arg *arg,
+		  uint16_t client_call_id)
+{
+	int i;
+	struct pptp_gre_con *gre_con;
+
+	for (i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++) {
+		gre_con = &arg->gre_cons[i];
+
+		if ((gre_con->flags & PPTP_ALG_FL_FREE_ENTRY) == 0 &&
+				  gre_con->ctx.client_call_id == client_call_id)
+			return gre_con;
+	}
+
+	return NULL;
+}
+
+/*
+ * PPTP TCP control connection ALG translator.
+ * It rewrites Call ID in the Outgoing-Call-Request
+ * message and Peer Call ID in the Outgoing-Call-Reply message.
+ */
+static bool
+npfa_pptp_tcp_translate(npf_cache_t *npc, npf_nat_t *nt, bool forw)
+{
+	uint16_t cksum;
+	uint16_t old_call_id;
+	uint32_t tcp_hdr_size;
+	uint32_t ip;
+	nbuf_t *nbuf;
+	struct tcphdr *tcp;
+	struct pptp_msg_hdr *pptp;
+	struct pptp_outgoing_call_reply *pptp_call_reply;
+	struct pptp_alg_arg *alg_arg;
+	struct pptp_gre_con *gre_con;
+	npf_cache_t gre_npc;
+	npf_addr_t *o_addr;
+	in_port_t o_port;
+
+	/* only ipv4 is supported so far */
+	if (!(npf_iscached(npc, NPC_IP4) && npf_iscached(npc, NPC_TCP) &&
+			  (npc->npc_l4.tcp->th_dport == htons(PPTP_SERVER_PORT) ||
+			  npc->npc_l4.tcp->th_sport == htons(PPTP_SERVER_PORT))))
+		return false;
+
+	nbuf = npc->npc_nbuf;
+	tcp = npc->npc_l4.tcp;
+	tcp_hdr_size = tcp->th_off << 2;
+	nbuf_reset(nbuf);
+
+	pptp = nbuf_advance(nbuf, npc->npc_hlen + tcp_hdr_size,
+			  sizeof(struct pptp_msg_hdr));
+	if (pptp == NULL)
+		return false;
+
+	if (pptp->pptp_msg_type != htons(PPTP_CTRL_MSG) ||
+			  pptp->len < htons(PPTP_OUTGOING_CALL_MIN_LEN) ||
+			  pptp->magic_cookie != htonl(PPTP_MAGIC_COOKIE))
+		return false;
+
+	/* get alg arg */
+	alg_arg = (struct pptp_alg_arg *)npf_nat_get_alg_arg(nt);
```

Comment:
I think the code here has concurrency problems: suppose there are two concurrent packets (let's say PPTP_OUTGOING_CALL_REQUEST) which could be processed on different CPUs. They might both attempt to initialise GRE connection list and/or allocate the GRE connection slot. In such case, these operations need to be serialised.

Reply:
yes, clearly I didn't give this code enough thoughts. I was thinking that RSS solves the concurrency problem for me by redirecting packets with the same src and dst ip addresses to the same queue/cpu. So, I can't rely on this in the kernel?

But even then there is race between incoming and outgoing pptp messages since they will be handled by different queues even if RSS is the case. I've added mutex to protect the handling of pptp messages. The contention is very unlikely since pptp control messages are very rare and it's even more unlikely that a user would establish several pptp tunnels to the same pptp server.

Also I've used a CAS to prevent double allocation of nat alg (gre connections slots).

Reply:
Using a lock fixes this.